File: Python/22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if n < 1:
            return [""]
        elif n == 1:
            return ["()"]
        
        current = set()
        current.add("()")

        for _ in range(2, n + 1):
            next = set()
            for s in current:
                for i in range(len(s)):
                    next.add(s[:i]+"()"+s[i:])
            current = next
        return list(current)

# @lc code=end

File: C++/22.generate-parentheses.cpp
/*
 * @lc app=leetcode id=22 lang=cpp
 *
 * [22] Generate Parentheses
 */

// @lc code=start
class Solution
{
public:
    void solve(string str, int op, int cl, vector<string> &sequences)
    {
        if (op == 0 && cl == 0)
        {
            sequences.push_back(str);
            return;
        }

        if (cl < op || op < 0 || cl < 0)
        {
            return;
        }

        solve(str + "(", op - 1, cl, sequences);
        solve(str + ")", op, cl - 1, sequences);
    }

    vector<string> generateParenthesis(int n)
    {
        vector<string> sequences = {};

        solve("", n, n, sequences);

        return sequences;
    }
};
// @lc code=end
